### PR TITLE
Stargaze Dockernet

### PR DESCRIPTION
**Pull request by sampocs**
Closes: #XXX

## Context and purpose of the change

* Added stargaze to dockernet to test before adding to mainnet. 
* Made the dockernet setup a little more flexible - so that we only have to change one variable to add/remove a host zone and we're not beholden to a specific ordering of registering host zones.  

## Brief Changelog
* Added stargaze to dockernet
* Added stars integration tests
* Updated hermes to the version that supports disabling batch transactions
* Updated the go relayer to support ICQs 
* Updated the dockernet build to insert a local address into the admin list
* Updated dockernet to enabled arbitrary ordering of host zones (previously that had to be registered in the same order or everything would break)
* Added IBC denom's across different channels to enable flexible host zone registration
* Updated the `create_logs` script to dynamically print logs for each host zone enabled 
* Removed `account_vars` script and defined functions in vars instead
* Added functions for determining validator and ICA addresses

## Author's Checklist
- Added Stars integration tests (ty @riley-stride )
- Ran all integration tests back to back
```
gaia_tests.bats
 ✓ [INTEGRATION-BASIC] address names are correct
 ✓ [INTEGRATION-BASIC] host zones successfully registered
 ✓ [INTEGRATION-BASIC-GAIA] ibc transfer updates all balances
 ✓ [INTEGRATION-BASIC-GAIA] liquid stake mints stATOM
 ✓ [INTEGRATION-BASIC-GAIA] tokens were transferred to GAIA after liquid staking
 ✓ [INTEGRATION-BASIC-GAIA] tokens on GAIA were staked
 ✓ [INTEGRATION-BASIC-GAIA] redemption works
 ✓ [INTEGRATION-BASIC-GAIA] claimed tokens are properly distributed
 ✓ [INTEGRATION-BASIC-GAIA] rewards are being reinvested, exchange rate updating
 ✓ [NOT-IMPLEMENTED] reinvestment and revenue amounts are correct

10 tests, 0 failures

juno_tests.bats
 ✓ [INTEGRATION-BASIC] address names are correct
 ✓ [INTEGRATION-BASIC] host zones successfully registered
 ✓ [INTEGRATION-BASIC-JUNO] ibc transfer updates all balances
 ✓ [INTEGRATION-BASIC-JUNO] liquid stake mints stJUNO
 ✓ [INTEGRATION-BASIC-JUNO] tokens were transferred to JUNO after liquid staking
 ✓ [INTEGRATION-BASIC-JUNO] tokens on JUNO were staked
 ✓ [INTEGRATION-BASIC-JUNO] redemption works
 ✓ [INTEGRATION-BASIC-JUNO] claimed tokens are properly distributed
 ✓ [INTEGRATION-BASIC-JUNO] rewards are being reinvested, exchange rate updating
 ✓ [NOT-IMPLEMENTED] reinvestment and revenue amounts are correct

10 tests, 0 failures

osmo_tests.bats
 ✓ [INTEGRATION-BASIC] address names are correct
 ✓ [INTEGRATION-BASIC] host zones successfully registered
 ✓ [INTEGRATION-BASIC-OSMO] ibc transfer updates all balances
 ✓ [INTEGRATION-BASIC-OSMO] liquid stake mints stOSMO
 ✓ [INTEGRATION-BASIC-OSMO] tokens were transferred to OSMO after liquid staking
 ✓ [INTEGRATION-BASIC-OSMO] tokens on OSMO were staked
 ✓ [INTEGRATION-BASIC-OSMO] redemption works
 ✓ [INTEGRATION-BASIC-OSMO] claimed tokens are properly distributed
 ✓ [INTEGRATION-BASIC-OSMO] rewards are being reinvested, exchange rate updating
 ✓ [NOT-IMPLEMENTED] reinvestment and revenue amounts are correct

10 tests, 0 failures

stars_tests.bats
 ✓ [INTEGRATION-BASIC] address names are correct
 ✓ [INTEGRATION-BASIC] host zones successfully registered
 ✓ [INTEGRATION-BASIC-STARS] ibc transfer updates all balances
 ✓ [INTEGRATION-BASIC-STARS] liquid stake mints stSTARS
 ✓ [INTEGRATION-BASIC-STARS] tokens were transferred to STARS after liquid staking
 ✓ [INTEGRATION-BASIC-STARS] tokens on STARS were staked
 ✓ [INTEGRATION-BASIC-STARS] redemption works
 ✓ [INTEGRATION-BASIC-STARS] claimed tokens are properly distributed
 ✓ [INTEGRATION-BASIC-STARS] rewards are being reinvested, exchange rate updating

9 tests, 0 failures
```

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes? 
  - [ ] Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`?
  - [ ] This pull request updates existing proto field values (and require a backend and frontend migration)? 
  - [ ] Does this pull request change existing proto field names (and require a frontend migration)?
  How is the feature or change documented? 
      - [ ] not applicable
      - [ ] jira ticket `XXX` 
      - [ ] specification (`x/<module>/spec/`) 
      - [X] README.md 
      - [ ] not documented

### CODE REVIEW

File: scripts/start_network.sh [hunk 1]
```diff
@@ -26,7 +26,7 @@ sh ${SCRIPT_DIR}/init_relayers.sh STRIDE ${HOST_CHAINS[@]}
 sh ${SCRIPT_DIR}/create_channels.sh ${HOST_CHAINS[@]}
 
 echo "Starting relayers"
-docker-compose up -d hermes icq
+docker-compose up -d hermes # icq
```
Comment (riley-stride) [inline scripts/start_network.sh]: nit should we just remove icq altogether? or worth keeping as we might need to bring it back if ibc-go breaks?

File: scripts-local/tests/gaia_tests.bats [hunk 2]
```diff
@@ -5,7 +5,7 @@ setup_file() {
   SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
   PATH="$SCRIPT_DIR/../../:$PATH"
 
-  #set allows us to export all variables in account_vars
+  # set allows us to export all variables in account_vars
```
Comment (riley-stride) [inline scripts-local/tests/gaia_tests.bats:8]: any idea why this line keeps getting uncommented?

File: scripts/tests/gaia_tests.bats [hunk 3]
```diff
@@ -58,29 +58,15 @@ setup() {
...
-  assert_equal $GAIA_ADDRESS "cosmos1pcag0cj4ttxg8l7pcg0q4ksuglswuuedcextl2"
-
-  assert_equal $DELEGATION_ICA_ADDR "cosmos1sy63lffevueudvvlvh2lf6s387xh9xq72n3fsy6n2gr5hm6u2szs2v0ujm"
-  assert_equal $REDEMPTION_ICA_ADDR "cosmos1xmcwu75s8v7s54k79390wc5gwtgkeqhvzegpj0nm2tdwacv47tmqg9ut30"
-  assert_equal $WITHDRAWAL_ICA_ADDR "cosmos1x5p8er7e2ne8l54tx33l560l8djuyapny55pksctuguzdc00dj7saqcw2l"
-  assert_equal $REVENUE_EOA_ADDR "cosmos1wdplq6qjh2xruc7qqagma9ya665q6qhcwju3ng"
-  assert_equal $FEE_ICA_ADDR "cosmos1lkgt5sfshn9shm7hd7chtytkq4mvwvswgmyl0hkacd4rmusu9wwq60cezx"
-  assert_equal $GAIA_DELEGATE_VAL_1 "cosmosvaloper1pcag0cj4ttxg8l7pcg0q4ksuglswuuedadj7ne"
-  assert_equal $GAIA_DELEGATE_VAL_2 "cosmosvaloper133lfs9gcpxqj6er3kx605e3v9lqp2pg5syhvsz"
+  assert_equal $(STRIDE_ADDRESS) "stride1uk4ze0x4nvh4fk0xm4jdud58eqn4yxhrt52vv7"
```
Comment (riley-stride) [inline scripts/tests/gaia_tests.bats:61]: just curious: why did you add the parens?

File: scripts/tests/gaia_tests.bats [hunk 4]
```diff
@@ -58,29 +58,15 @@ setup() {
...
+  assert_equal $(GAIA_ADDRESS) "cosmos1pcag0cj4ttxg8l7pcg0q4ksuglswuuedcextl2"
+
+  assert_equal $GAIA_DELEGATION_ICA_ADDR "cosmos1sy63lffevueudvvlvh2lf6s387xh9xq72n3fsy6n2gr5hm6u2szs2v0ujm"
+  assert_equal $GAIA_REDEMPTION_ICA_ADDR "cosmos1xmcwu75s8v7s54k79390wc5gwtgkeqhvzegpj0nm2tdwacv47tmqg9ut30"
+  assert_equal $GAIA_WITHDRAWAL_ICA_ADDR "cosmos1x5p8er7e2ne8l54tx33l560l8djuyapny55pksctuguzdc00dj7saqcw2l"
+  assert_equal $GAIA_FEE_ICA_ADDR "cosmos1lkgt5sfshn9shm7hd7chtytkq4mvwvswgmyl0hkacd4rmusu9wwq60cezx"
+  assert_equal $GAIA_DELEGATE_VAL "cosmosvaloper1pcag0cj4ttxg8l7pcg0q4ksuglswuuedadj7ne"
   assert_equal $GAIA_RECEIVER_ACCT "cosmos1g6qdx6kdhpf000afvvpte7hp0vnpzapuyxp8uf"
```
Comment (riley-stride) [inline scripts/tests/gaia_tests.bats:69]: Thanks for cleaning this up. 

Noticing that the int tests across chains are sufficiently similar that I think we could probably write a single well-paramaterized `chain_int_tests.bats` and pass in the chain names we'd like to test in the same way as we do for build. For example, we could call `make test-integration chains=gsj` to test gaia stargaze and juno, but not osmosis. 

Don't think this is worth doing now. Maybe after the next chain?

File: scripts/vars.sh [hunk 5]
```diff
@@ -280,3 +280,18 @@ WAIT_FOR_BLOCK() {
...
+
+GET_VAL_ADDR() {
+  chain_id=$1
+  val_index=$2
+
+  MAIN_CMD=$(GET_VAR_VALUE ${chain_id}_MAIN_CMD)
+  $MAIN_CMD q staking validators | grep ${chain_id}_${val_index} -A 5 | grep operator | awk '{print $2}'
+}
+
+GET_ICA_ADDR() {
```
Comment (riley-stride) [inline scripts/vars.sh:295]: very nice

File: scripts/tests/stars_tests.bats [hunk 6]
```diff
@@ -60,11 +60,11 @@ setup() {
 @test "[INTEGRATION-BASIC] address names are correct" {
   assert_equal $(STRIDE_ADDRESS) "stride1uk4ze0x4nvh4fk0xm4jdud58eqn4yxhrt52vv7"
 
-  assert_equal $STARS_DELEGATE_VAL 'starsvaloper12ffkl30v0ghtyaezvedazquhtsf4q5ng2c0xaf'
-  assert_equal $STARS_DELEGATION_ICA_ADDR "stars1kl6wa99e6hf97xr90m2n04rl0smv842pj9utqyvgyrksrm9aacdqyfc3en"
-  assert_equal $STARS_REDEMPTION_ICA_ADDR "stars1x07hv0hxujj6l0mfyynwyuccf8fl27vjup0y8dmyajy9ugae22hqfvmv4e"
-  assert_equal $STARS_WITHDRAWAL_ICA_ADDR "stars1x5ndl5p9tjy376a9xmqhw79gz0s678480759cdgaretcgm36akvs0a78tj"
-  assert_equal $STARS_FEE_ICA_ADDR "stars1v09y993sku5djvm0rffq0nfsk5rzke4d2vzvny5e6vmq7dz0dehqnwl4ay"
+  assert_equal $(GET_VAL_ADDR STARS 1) 'starsvaloper12ffkl30v0ghtyaezvedazquhtsf4q5ng2c0xaf'
```
Comment (riley-stride) [inline scripts/tests/stars_tests.bats:63]: nice step toward parameterizing the int tests

File: scripts/vars.sh [hunk 7]
```diff
@@ -220,6 +220,7 @@ GAIA_REDEMPTION_ICA_ADDR='cosmos1xmcwu75s8v7s54k79390wc5gwtgkeqhvzegpj0nm2tdwacv
 GAIA_WITHDRAWAL_ICA_ADDR='cosmos1x5p8er7e2ne8l54tx33l560l8djuyapny55pksctuguzdc00dj7saqcw2l'
 GAIA_FEE_ICA_ADDR='cosmos1lkgt5sfshn9shm7hd7chtytkq4mvwvswgmyl0hkacd4rmusu9wwq60cezx'
 GAIA_RECEIVER_ACCT='cosmos1g6qdx6kdhpf000afvvpte7hp0vnpzapuyxp8uf'
+GAIA_REVENUE_ACCT=''
```
Comment (riley-stride) [inline scripts/vars.sh]: nit: why not rm?

File: scripts/test-util/1.sh [hunk 8]
```diff
@@ -1,8 +1,8 @@
 ### IBC TRANSFER
 SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
```
Comment (riley-stride) [inline scripts/test-util/1.sh:2]: If you intended to commit the test-util/N.sh scripts, could you please add a comment in the PR description as to why?

**Review by riley-stride — APPROVED** [review]
Your diligence in cleaning up technical debt here is going to pay some massive dividends when we onboard the next host zone. Looks great. Only some very minor comments.

File: scripts/start_network.sh (hunk 1 above)
Comment (sampocs) [reply scripts/start_network.sh]: I think this file might be out of date - its removed in the latest commit

File: scripts-local/tests/gaia_tests.bats (hunk 2 above)
Comment (sampocs) [reply scripts-local/tests/gaia_tests.bats:8]: not a clue, but I'll keep an eye on it - this is the first time I noticed it

File: scripts/tests/gaia_tests.bats (hunk 3 above)
Comment (sampocs) [reply scripts/tests/gaia_tests.bats:61]: Good question! 

So before each host address was defined as a variable like so:
```bash
STRIDE_ADDRESS=$(...)
GAIA_ADDRESS=$(...)
OSMO_ADDRESS=$(...)
etc...
```
Issue with that was when you'd source the file to load each variable, it would try to run each line, and if one of the host zones was not currently running, the script would error. So previously if, for instance, you wanted to start the network with just stride and gaia, you'd have to comment out the OSMO_ADDRESS line, otherwise the startup would fail.

I updated it so that now instead of a variable, it's a function
```
function STRIDE_ADDRESS() {...}
function GAIA_ADDRESS() {...}
function OSMO_ADDRESS() {...}
etc...
```
So now when you source the file, it simply loads the functions - but it doesn't run the code that's inside the function. And the code inside the function isn't run until you explicitly call it with $(STRIDE_ADDRESS)

Variable Access: ${VAR_NAME} or $VAR_NAME 
Function Call: $(FUNCTION_NAME)

Other benefit of this is that now there's no need to have a separate account_var script

File: scripts/tests/gaia_tests.bats (hunk 4 above)
Comment (sampocs) [reply scripts/tests/gaia_tests.bats:69]: Yup I was thinking the same thing! Also I got a slight head start on this if you look at the [stars integration tests](https://github.com/Stride-Labs/stride/blob/stars-dockernet/scripts/tests/stars_tests.bats#L63-L67) where the validator and ICA addresses are generated dynamically. 

I didn't want to go through the headache of updating GAIA,OSMO,JUNO atm since it would have probably warranted a couple more hours of running tests haha but planning to do it when I add the next host. 

And it's still a little unclear to me how involved the full refactor would be (i.e. one single parameterized script) - but I'll re-evaluate as I add the next host!

File: scripts/vars.sh (hunk 7 above)
Comment (sampocs) [reply scripts/vars.sh]: removed in latest commit - I wasn't sure what this was used for at the time but I later realized it was just an alias of the fee ICA and removed it

File: scripts/test-util/1.sh (hunk 8 above)
Comment (sampocs) [reply scripts/test-util/1.sh:2]: Discussed offline - going to keep scripts 1-4 but remove scratch files
